source/main.cpp

```diff
@@ -18,156 +18,266 @@
+            const u8 *logo = logo_bin;
+
+            for (s32 y1 = 0; y1 < 31; y1++) {
+                for (s32 x1 = 0; x1 < 84; x1++) {
+                    const tsl::gfx::Color color = {static_cast<u8>(logo[3] >> 4), static_cast<u8>(logo[2] >> 4), static_cast<u8>(logo[1] >> 4), static_cast<u8>(logo[0] >> 4)};
+                    renderer->setPixelBlendSrc(20 + x1, 20 + y1, renderer->a(color));
+                    logo += 4;
+                }
+            }
+
+            renderer->drawString(envGetLoaderInfo(), false, 20, 68, 15, renderer->a(tsl::style::color::ColorText));
+        }));
+
+        this->refreshGui();
 
         return rootFrame;
     }
+
+    void refreshGui() {
+        this->removeFocus();
+        auto browser = std::make_unique<FileBrowser>();
+        Result rc = browser->scanCwd();
+
+        if (R_FAILED(rc)) {
+            /* Something went wrong. */
+            char resultBuffer[0x10];
+            std::snprintf(resultBuffer, 0x10, "2%03X-%04X", R_MODULE(rc), R_DESCRIPTION(rc));
+            rootFrame->setContent(new tsl::elm::CustomDrawer([resultBuffer](tsl::gfx::Renderer *renderer, u16 x, u16 y, u16 w, u16 h) {
+                renderer->drawString("\uE150", false, (tsl::cfg::FramebufferWidth - 90) / 2, 300, 90, renderer->a(tsl::style::color::ColorText));
+                renderer->drawString("Failed to scan cwd!", false, 105, 380, 25, renderer->a(tsl::style::color::ColorText));
+                renderer->drawString(resultBuffer, false, 82, 410, 15, renderer->a(tsl::style::color::ColorDescription));
+            }));
+        } else if (browser->isEmpty()) {
+            /* Folder is empty */
+            rootFrame->setContent(new tsl::elm::CustomDrawer([](tsl::gfx::Renderer *renderer, u16 x, u16 y, u16 w, u16 h) {
+                renderer->drawString("\uE150", false, (tsl::cfg::FramebufferWidth - 90) / 2, 300, 90, renderer->a(tsl::style::color::ColorText));
+                renderer->drawString("No Overlays found!", false, 105, 380, 25, renderer->a(tsl::style::color::ColorText));
+                renderer->drawString("Place your .ovl files in /switch/.overlays", false, 82, 410, 15, renderer->a(tsl::style::color::ColorDescription));
+            }));
+        } else {
+            /* Release actual browser gui */
+            rootFrame->setContent(browser.release());
+        }
+    }
 };
 
 class OverlayTeslaMenu : public tsl::Overlay {
-public:
-    OverlayTeslaMenu() { }
-    ~OverlayTeslaMenu() { }
-
-    void onShow() override { 
-        if (rootFrame != nullptr) {
-            tsl::Overlay::get()->getCurrentGui()->removeFocus();
-            rebuildUI();
-            rootFrame->invalidate();
-            tsl::Overlay::get()->getCurrentGui()->requestFocus(rootFrame, tsl::FocusDirection::None);
-        }
+  public:
+    OverlayTeslaMenu() {}
+    ~OverlayTeslaMenu() {}
+
+    void onShow() override {
+        auto *gui = dynamic_cast<GuiMain *>(tsl::Overlay::get()->getCurrentGui().get());
```

Comment:
This technically isn't necessary right now but I guess in the future if there are more Guis?